spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,322 @@
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-defineownproperty-p-desc">
+        <h1>[[DefineOwnProperty]] ( _P_, _Desc_ )</h1>
+        <p>When the [[DefineOwnProperty]] internal method of a Tuple exotic object _T_ is called with property key _P_, and Property Descriptor _Desc_ the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If _Desc_.[[Writable]] is present and has value *true*, return *false*.
+          1. If _Desc_.[[Enumerable]] is present and has value *false*, return *false*.
+          1. If _Desc_.[[Configurable]] is present and has value *true*, return *false*.
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Let _current_ be ? TupleGet(_T_, _numericIndex_)
+              1. If _current_ is ~empty~ return *false*.
+              1. If IsAccessorDescriptor(_Desc_) is *true*, return *false*.
+              1. If _Desc_.[[Value]] is present, return ! SameValue(_Desc_.[[Value]], _current_).
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+            
+      <emu-clause id="sec-tuple-exotic-objects-get-p-receiver">
+        <h1>[[Get]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Get]] internal method of a Tuple exotic object _T_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Let _value_ be ? TupleGet(_T_, _numericIndex_).
+              1. If _value_ is ~empty~, return *undefined*.
+              2. Return _value_.
+          1. Let _parent_ be ? _T_.[[GetPrototypeOf]]().
+          1. Return ? _parent_.[[Get]](_P_, _Receiver_).
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-set-p-receiver">
+        <h1>[[Set]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Set]] internal method of a Tuple exotic object _T_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-tuple-exotic-objects-ownpropertykeys">
+        <h1>[[OwnPropertyKeys]] ( )</h1>
+        <p>When the [[OwnPropertyKeys]] internal method of a Record exotic object _R_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Assert _T_ is Tuple exotic object.
+          1. Let _len_ be the length of _T_.[[TupleData]].[[Sequence]].
+          1. For each integer _i_ starting with 0 such that _i_ &lt; _len_ in ascending order, do
```

Comment:
this reads awkwardly to me; usually this is done with an i variable, set to 0, and a while loop.

Reply:
I can change that but FYI, you might consider an editorial change here: https://tc39.es/ecma262/#sec-string-exotic-objects-ownpropertykeys as I just took the same formulation